### PR TITLE
[hw/ip] OTP ROM patch generation tool

### PR DESCRIPTION
**Pull request by sameo**
This PR is a first attempt at building an OTP ROM patch generation tool, based on the proposed ROM patch specification changes tracked at #21070 

TODO before this PR can be merged:

* [x] Add tests to validate the generated patches
* [x] Define multiple output (OTP HJSON partition, binary blob)
* [x] Add ROM patch signature support

### CODE REVIEW

File: third_party/rust/Cargo.toml [hunk 1]
```diff
@@ -55,8 +55,8 @@ rsa = "0.8"
 rusb = "0.8.1"
 serde = { version="1.0", features=["serde_derive"] }
 serde_derive = "1.0"
-
 serde_bytes = "0.11"
+serde-hjson = "0.9"
```
Comment (engdoreis) [inline third_party/rust/Cargo.toml:59]: @pamaury have more context on this, but I think that we had issues trying to parse some hjson files in the repo with the rust implementation. So unless we really need the hjson features, here I would suggest using json.

File: sw/host/opentitantool/src/command/rom_patch.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    fn run(
+        &self,
+        _context: &dyn Any,
+        _transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        let mut patch_partition = RomPatchPartition::new(&self.patch, &self.key)?;
+        let mut buffer = vec![0_u8; self.otp_size];
+        _ = patch_partition.read(&mut buffer)?;
+
+        println!("uint8_t fake_patch[] =");
```
Comment (engdoreis) [inline sw/host/opentitantool/src/command/rom_patch.rs:56]: We have other similar tooling in the repo to generate code from a template + hjson file using the python library [mako templates](https://www.makotemplates.org/).
Is there a reason for the tool to be inside opentitantool or can this also be a python script using a template file?
In case it needs to be inside opentitantool, can we use a template crate?

**Review by engdoreis — COMMENTED** [review]
Thanks @sameo for the PR, I please check the questions left below.

**Comment by andreaskurth** [comment]
Thanks for this work and PR, @sameo! Let's get more reviews on this, so we can eventually be confident with merging it :-)

I'd like to review the first part you mention, "A ROM patch specification change", but couldn't find it in the files changed in this PR. Could you please point me to it?

File: third_party/rust/Cargo.toml (hunk 1 above)
Comment (jwnrt) [reply third_party/rust/Cargo.toml:59]: @engdoreis is right, but if you _do_ need to use hjson then we found the [deser-hjson](https://crates.io/crates/deser-hjson) crate to have fewer bugs (but not zero) than serde-hjson.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO(lowRISC/opentitan#15443): Fix this lint.
+#[allow(clippy::module_inception)]
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/mod.rs:6]: I think you can remove these lines - the parent module is named `rom_patch` and the child is just `patch`, so there shouldn't be a warning.

Also, nit: do you need this `patch` submodule? Are you planning to add more modules to `rom_patch` in the future?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+use memoffset::offset_of;
+use serde::Deserialize;
+
+use std::fmt;
+use std::fs;
+use std::io::{self, Read};
+use std::ops::Range;
+use std::path::Path;
+
+use std::{mem, ptr};
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:18]: Nit:

```suggestion
use std::path::Path;
use std::mem;
use std::ptr;
```

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+    N_MATCH_DESCRIPTORS * mem::size_of::<RomPatchMatchDescriptor>();
+
+// The code section offset from the patch start
+const PATCH_CODE_SECTION_OFFSET: usize = mem::size_of::<RomPatchHeader>() + PATCH_MATCH_TABLE_SIZE;
+
+#[allow(dead_code)]
+enum MultiBitBool {
+    Bool4True = 0x6,
+    Bool4False = 0x9,
+}
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:44]: Nit:

These constants are already available from [`crate::chip::boolean`](https://opentitan.org/gen/rustdoc/opentitanlib/chip/boolean/struct.MultiBitBool4.html) as `MultiBitBool4::True as u8` and `MultiBitBool4::False as u8`.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        unsafe {
+            ptr::copy_nonoverlapping(raw_header, buf.as_mut_ptr(), header_size);
+        }
+
+        Ok(header_size)
+    }
+}
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:154]: Per @nbdd0121 - this probably shouldn't be an unsafe `Read` implementation, you can use [`zerocopy::AsBytes`](https://docs.rs/zerocopy/0.7.32/zerocopy/trait.AsBytes.html) instead. This is already in our dependency tree.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+            "\tPatch size {:?} Locked {:?} ",
+            self.size(),
+            self.locked()
+        )
+    }
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+#[allow(dead_code)]
+#[repr(packed)]
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:132]: This needs to be `#[repr(packed, C)]` otherwise it will be `#[repr(packed, Rust)]` and rust can re-order the fields.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 8]
```diff
@@ -0,0 +1,738 @@
...
+
+impl fmt::Display for RomPatchHeader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let unaligned = std::ptr::addr_of!(self.size);
+        let size = unsafe { std::ptr::read_unaligned(unaligned) };
+        write!(
+            f,
+            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
+            self.meta, self.revision, size
+        )
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:164]:
```suggestion
        write!(
            f,
            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
            self.meta, self.revision, {self.size}
        )
```

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 9]
```diff
@@ -0,0 +1,738 @@
...
+        let range = self.range();
+        write!(
+            f,
+            "\tBase address {:#04x} Range [{:#04x}:{:#04x}]",
+            self.base_addr, range.start, range.end
+        )
+    }
+}
+
+impl Read for RomPatchSubRoutine {
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:227]: This can also be a `zerocopy::AsBytes` call I think

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 10]
```diff
@@ -0,0 +1,738 @@
...
+            .to_le_bytes()[..];
+
+        /* Copy the signature */
+        unsafe {
+            ptr::copy_nonoverlapping(
+                signature.as_ptr(),
+                buf[self.signature_offset()..].as_mut_ptr(),
+                RSA_SIGNATURE_LEN_BYTES,
+            );
+        }
```
Comment (jwnrt) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:444]: `signature` is already a `&[u8]` so I don't think you need to copy the memory like this.

**Review by jwnrt — CHANGES_REQUESTED** [review]
Thanks @sameo - some comments on the unsafe code

**Review by alees24 — COMMENTED** [review]
Thanks for this. The documentation/specification changes look good, modulo a few corrections and clarifications that I've suggested.
May I suggest, given the time frames involved, that it would be sensible to separate these from the patch generation tool so that the documentation and specification changes may be integrated sooner?

The lowRISC software team will want to review the tool and Rust code thoroughly and I think our priority should be to get the ROMs and e2e testing completed first.

File: third_party/rust/Cargo.toml (hunk 1 above)
Comment (pamaury) [reply third_party/rust/Cargo.toml:59]: Unless we want to serialize hjson, I would indeed suggest that we stick `deser-hjson` which we already use and which is working well (we successfully submitted upstream bug fixes).

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+// The code section offset from the patch start
+const PATCH_CODE_SECTION_OFFSET: usize = mem::size_of::<RomPatchHeader>() + PATCH_MATCH_TABLE_SIZE;
+
+#[allow(dead_code)]
+enum MultiBitBool {
+    Bool4True = 0x6,
+    Bool4False = 0x9,
+}
+
+#[derive(Deserialize, Debug)]
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:46]: Unless I have missed it, the hjson format is not really documented. There is a (quite clear) example but I think it would be worth either having a proper document explaining the format, or having documentation comments in the rust code.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 12]
```diff
@@ -0,0 +1,745 @@
...
+pub struct RomPatchPartitionHjson {
+    patches: Vec<RomPatchHjson>,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(dead_code)]
+#[repr(packed)]
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:74]: Should this be `repr(packed, C)`?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+    regions: Vec<RomPatchRegionHjson>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct RomPatchPartitionHjson {
+    patches: Vec<RomPatchHjson>,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(dead_code)]
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:73]: Why is this necessary?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+struct RomPatchMatchDescriptor {
+    match_base: u32,
+    remap_base: u32,
+}
+
+impl RomPatchMatchDescriptor {
+    fn from_hjson(remap_base: u32, region: &RomPatchRegionHjson) -> Result<Self> {
+        let match_addr = u32::from_str(&region.match_addr)?;
+
+        let size_dwords = (region.patch_code.len() as u32) & 0xf;
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:84]: should the code make sure that `patch_code` is not too long?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+        let locked = if region.locked { 1 << 31 } else { 0 };
+
+        let match_address = match_addr & 0x7ffffff;
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:87]: what happens if match_addr has its high bit set? Is that an error?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+
+    fn remap_addr(&self) -> u32 {
+        self.remap_base
+    }
+
+    fn locked(&self) -> bool {
+        (self.match_base >> 31) != 0
+    }
+
+    fn size(&self) -> u8 {
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:108]: I suggest renaming `size` to something more precise, like `size_dwords` (which I think is what it is here?)

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 17]
```diff
@@ -0,0 +1,745 @@
...
+        let offset = PATCH_CODE_SECTION_OFFSET + (base_addr - patch_base_addr) as usize;
+
+        Ok(Self {
+            base_addr,
+            offset,
+            code,
+        })
+    }
+
+    fn len(&self) -> usize {
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:200]: I would suggest to rename `len` to something more precise, like `size_bytes`.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 18]
```diff
@@ -0,0 +1,745 @@
...
+            buf_offset += 4;
+        }
+
+        Ok(self.len())
+    }
+}
+
+struct RomPatch<'a> {
+    header: RomPatchHeader,
+    size: usize,
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:250]: maybe document what this is the size of? And size in what unit (bytes?)

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 19]
```diff
@@ -0,0 +1,745 @@
...
+        let base_addr = u32::from_str(&patch.base_addr)?;
+        let mut remap_addr = base_addr;
+        let mut match_table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS] = Default::default();
+        let mut patch_code = vec![];
+        let mut sub_routines = vec![];
+
+        for (index, region) in patch.regions.iter().enumerate() {
+            let match_pair = RomPatchMatchDescriptor::from_hjson(remap_addr, region)?;
+            match_table[index] = match_pair;
+            remap_addr += (region.patch_code.len() as u32) * 4;
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:272]: replace 4 by `std::size_of<u32>()`? Apply throughout.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 20]
```diff
@@ -0,0 +1,745 @@
...
+
+    fn code_offset(&self) -> usize {
+        self.code_offset
+    }
+
+    fn signature_offset(&self) -> usize {
+        self.signature_offset
+    }
+
+    fn code_len(&self) -> usize {
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:325]: rename to specify the unit

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 21]
```diff
@@ -0,0 +1,745 @@
...
+        self.signature_offset() - self.code_offset()
+    }
+
+    fn len(&self) -> usize {
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:329]: what is this the length of?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 22]
```diff
@@ -0,0 +1,745 @@
...
+        self.size
+    }
+
+    fn validate(&self) -> Result<()> {
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:333]: what is the purpose of this method?

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 23]
```diff
@@ -0,0 +1,745 @@
...
+            "\tDescriptors {:?} bytes",
+            mem::size_of_val(&self.match_table)
+        )?;
+        writeln!(f, "\tCode {:?} bytes", self.code_len())?;
+        writeln!(f, "\tSignature {:?}", RSA_SIGNATURE_LEN_BYTES)
+    }
+}
+
+impl<'a> Read for RomPatch<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:370]: As @jwnrt said above, this could use `zerobytes::FromBytes` to make it safer and clearer. See an example here:
https://github.com/lowRISC/opentitan/pull/20672/files#diff-4455f74af62f7ea7ec84c7fd2e934a0b14e940e9d917ae899d276fc5970968afR315

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 24]
```diff
@@ -0,0 +1,745 @@
...
+                RSA_SIGNATURE_LEN_BYTES,
+            );
+        }
+
+        Ok(self.len())
+    }
+}
+
+pub struct RomPatchPartition<'a> {
+    size: usize,
```
Comment (pamaury) [inline sw/host/opentitanlib/src/rom_patch/patch.rs:451]: size of what?

File: sw/host/opentitantool/src/command/rom_patch.rs [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+
+use opentitanlib::app::command::CommandDispatch;
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::crypto::rsa::RsaPrivateKey;
+use opentitanlib::rom_patch::patch::RomPatchPartition;
+
+/// Generate an OTP ROM patch partition
+#[derive(Debug, StructOpt)]
+pub struct Generate {
+    #[structopt(
```
Comment (pamaury) [inline sw/host/opentitantool/src/command/rom_patch.rs:21]: Please use `clap`, we migrated away from `structopt` (which is the predecessor of structopt)

File: sw/host/opentitantool/src/command/rom_patch.rs [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+        short = "s",
+        long = "otp_size",
+        name = "OTP_SIZE",
+        help = "OTP ROM patch partition total size (bytes)"
+    )]
+    otp_size: usize,
+}
+
+impl CommandDispatch for Generate {
+    fn run(
```
Comment (pamaury) [inline sw/host/opentitantool/src/command/rom_patch.rs:47]: Can you explain/document what this command is supposed to generate? I was expecting it to generate a binary file (i.e. one that corresponds to the format that you implemented `Read` for). But this seems to generate a C file. What will this C file be used for?

**Comment by sameo** [comment]
> Thanks for this. 

Thanks for the review.

> The documentation/specification changes look good, modulo a few corrections and clarifications that I've suggested. May I suggest, given the time frames involved, that it would be sensible to separate these from the patch generation tool so that the documentation and specification changes may be integrated sooner?
> 

Yes, that makes sense. I have now extracted and removed the specification changes from this PR into #21070 so that we can merge it first.

> The lowRISC software team will want to review the tool and Rust code thoroughly and I think our priority should be to get the ROMs and e2e testing completed first.

I agree. There are quite a lot of pending comments on the rust tool in this PR, I will address them very shortly.

File: third_party/rust/Cargo.toml (hunk 1 above)
Comment (sameo) [reply third_party/rust/Cargo.toml]: This is fixed with the updated PR.

File: sw/host/opentitantool/src/command/rom_patch.rs (hunk 2 above)
Comment (sameo) [reply sw/host/opentitantool/src/command/rom_patch.rs]: We are now generating an HJSON OTP partition image, to be used as an OTP image overlay.

File: sw/host/opentitanlib/src/rom_patch/mod.rs (hunk 3 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: I removed that comment. And I also removed `patch.rs` to keep all the code under `mod.rs`.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 4 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Fixed as well, thanks.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 5 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Yep. I switched to the internal `MultiBitBool4` crate with the new PR.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 6 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: All the code has been moved to use the `zerocopy` crate.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 7 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Fixed as well, thanks.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 8 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: This part of the code has been reworked. We should not do unaligned accesses anymore, except in the unit tests.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 10 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: I switched to ECDSA signatures, and that has been fixed now.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 9 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Indeed. We no longer use the `Read` trait.

File: hw/ip/rv_core_ibex/doc/boot-rom-patching.md [hunk 27]
```diff
@@ -140,7 +140,7 @@ The second ROM partition with patching capability is added to the architecture d
 
 - **The patch table** includes:
 
-  - Contents for 32 [*patch match register pairs*](#patch-match--redirect-logic), each register 1 DWORD in size.
+  - Contents for 32 [*patch match descriptor pairs*](#patch-match--redirect-logic), each descriptor is 1 DWORD in size.
```
[the remark this answers is not in the document]
Comment (sameo) [reply hw/ip/rv_core_ibex/doc/boot-rom-patching.md]: Closing all those comments as the related documentation changes have been merged in a separate PR.
Thanks for the comments @alees24.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 13 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Not relevant anymore with the update PR.

File: sw/host/opentitantool/src/command/rom_patch.rs (hunk 25 above)
Comment (sameo) [reply sw/host/opentitantool/src/command/rom_patch.rs]: I had to pull the `clap` crate update and the migration code, in order to proceed with that conversion.

File: sw/host/opentitantool/src/command/rom_patch.rs (hunk 26 above)
Comment (sameo) [reply sw/host/opentitantool/src/command/rom_patch.rs:61]: We now generate a blob into a file, or an HJSON OTP image partition to be used as an OTP image overlay. I hope it makes more sense.

File: sw/host/opentitantool/src/command/rom_patch.rs (hunk 25 above)
Comment (sameo) [reply sw/host/opentitantool/src/command/rom_patch.rs]: Done.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 16 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Not relevant anymore with the updated PR.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 17 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: ditto.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 18 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Removed from the updated PR.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 19 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Yes. I have done that throughout all the code.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 20 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Not relevant anymore.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 21 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Ditto.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 23 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Yes. The updated PR is now based on the `zerocopy` crate.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 24 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Not relevant anymore with the updated PR.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 22 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: I removed it.

File: sw/host/opentitanlib/src/rom_patch/patch.rs [hunk 28]
```diff
@@ -0,0 +1,745 @@
...
+        for s in &sub_routines {
+            code_addr_end = std::cmp::max(code_addr_end, s.range().end);
+        }
+
+        let code_len = (code_addr_end - base_addr) as usize;
+        let patch_size = mem::size_of::<RomPatchHeader>()
+            + (N_MATCH_DESCRIPTORS * mem::size_of::<RomPatchMatchDescriptor>())
+            + code_len
+            + RSA_SIGNATURE_LEN_BYTES;
+
```
[the remark this answers is not in the document]
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: Good point. I added a check for that.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 15 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: I added a check for that. It should be an error indeed.

File: sw/host/opentitanlib/src/rom_patch/patch.rs (hunk 14 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/patch.rs]: That part has been removed from the PR.

**Comment by sameo** [comment]
@pamaury @engdoreis @jwnrt I finally found some bandwidth to fixed that PR. The forcing function was the ECDSA enablement in `integrated_dev`...

Anyways, this updated PR addresses all pending comments, hopefully. Using the `zerocopy` crate definitely makes the code smaller and there are no unsafe blocks anymore.

As I also wanted to switch to `clap` for the CLI and to ECDSA for the ROM patch signature, this PR carries some commits from `master`.

Thanks in advance for giving that PR a new round of reviews.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 29]
```diff
@@ -0,0 +1,691 @@
...
+
+        // Compute the overall code size for the patch.
+        for region in regions {
+            if let Some(sub_routines_hjson) = &region.sub_routines {
+                for s in sub_routines_hjson {
+                    code_size_words =
+                        std::cmp::max(code_size_words, (s.offset >> 2) + s.code.len());
+                }
+            }
+        }
```
Comment (engdoreis) [inline sw/host/opentitanlib/src/rom_patch/mod.rs:181]: You could remove the Option from `sub_routines` combined with `#[serde(default)]` to make it an empty vector in case it is not defined in the hjson, so that the `if let` would not be required. 
Iterator functions could also be used to make the code more idiomatic.
```suggestion
let code_size_words = regions
        .iter()
        .flat_map(|region| region.sub_routines.iter())
        .map(|s| (s.offset >> 2) + s.code.len())
        .fold(code_size_words, std::cmp::max);
```

**Review by engdoreis — COMMENTED** [review]
Thanks @sameo for the PR.

I've reviewed most of it and let some comments.
I'm going to finish the review of the last commit later.

**Review by jesultra — APPROVED** [review]
I do not know about "ROM patches", but the switch to clap for command line parsing, and general structure of the code look fine to me.

File: sw/host/opentitanlib/src/otp/otp_img.rs [hunk 30]
```diff
@@ -31,6 +33,19 @@ where
...
 
+fn serialize_vec_u32_hex<S>(vec: &[u32], serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    let strings: Vec<String> = vec.iter().map(|&num| format!("0x{:08x}", num)).collect();
+    let mut seq = serializer.serialize_seq(Some(strings.len()))?;
+    for s in &strings {
+        seq.serialize_element(s)?;
+    }
```
[the remark this answers is not in the document]
Comment (sameo) [reply sw/host/opentitanlib/src/otp/otp_img.rs]: Thanks, suggestion applied now.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 31]
```diff
@@ -0,0 +1,691 @@
...
+
+        let size_dwords = match patch_code_size_words {
+            1 | 2 | 4 | 8 => (patch_code_size_words as u32) & PATCH_SIZE_FIELD_MASK,
+            _ => bail!("Invalid ROM patch region code size {patch_code_size_words}"),
+        };
+        let locked = if region.locked {
+            1 << PATCH_LOCKED_FIELD_SHIFT
+        } else {
+            0
+        };
```
[the remark this answers is not in the document]
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: Applied as well, thanks.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 32]
```diff
@@ -0,0 +1,691 @@
...
+
+        let match_address = match match_addr {
+            addr if addr <= PATCH_MATCH_ADDR_FIELD_MASK => addr & PATCH_MATCH_ADDR_FIELD_MASK,
```
[the remark this answers is not in the document]
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: Indeed. Fixed now.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 33]
```diff
@@ -0,0 +1,691 @@
...
+            match_base,
+            remap_base,
+        })
+    }
+}
+
+#[derive(AsBytes, Copy, Clone, Debug, Default)]
+#[repr(packed, C)]
+struct RomPatchMatchTable {
+    table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS],
```
[the remark this answers is not in the document]
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: N_MATCH_DESCRIPTORS is an RTL constant for Darjeeling. But I guess using const generics make it more flexible. I switched the implementation to it now, and although rom_patch.rs uses the exported N_MATCH_DESCRIPTORS, this could become a CLI option in the future.

File: sw/host/opentitanlib/src/rom_patch/mod.rs (hunk 29 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: Nice. Applied, thanks.

**Comment by sameo** [comment]
> Thanks @sameo for the PR.
> 
> I've reviewed most of it and let some comments. I'm going to finish the review of the last commit later.

Thanks for the review @engdoreis . I think I addressed all your comments now.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 34]
```diff
@@ -0,0 +1,700 @@
...
+
+    fn len_bytes(&self) -> usize {
+        self.code.len() * mem::size_of::<u32>()
+    }
+
+    fn as_bytes(&self) -> &[u8] {
+        unsafe {
+            let ptr = self.code.as_ptr();
+            let len = self.code.len() * std::mem::size_of::<u32>();
+            std::slice::from_raw_parts(ptr as *const u8, len)
```
Comment (engdoreis) [inline sw/host/opentitanlib/src/rom_patch/mod.rs:215]: Only this line needs to be unsafe.

File: sw/host/opentitantool/src/command/rom_patch.rs [hunk 35]
```diff
@@ -0,0 +1,106 @@
...
+                    .as_bytes(),
+            )?;
+            Ok(None)
+        } else {
+            Ok(Some(Box::new(otp_img)))
+        }
+
+        // let json = serialize(&otp_img)?.to_hjson().to_string();
+
+        // if let Some(hjson_output) = &self.hjson_output {
```
Comment (engdoreis) [inline sw/host/opentitantool/src/command/rom_patch.rs:89]: Remove commented out code?

**Review by engdoreis — APPROVED** [review]
Thanks @sameo for the fixes, besides some other comments below it LGTM.

File: sw/host/opentitanlib/src/rom_patch/mod.rs [hunk 36]
```diff
@@ -0,0 +1,700 @@
...
+            .iter()
+            .flat_map(|region| region.sub_routines.iter())
+            .map(|s| (s.offset >> 2) + s.code.len())
+            .fold(code_size_words, std::cmp::max);
+
+        let mut code = vec![RV32_NOP; code_size_words];
+        let mut code_index = 0;
+
+        // Copy the patch code itself.
+        for region in regions {
```
[the remark this answers is not in the document]
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: Sure. Done.

File: sw/host/opentitantool/src/command/rom_patch.rs (hunk 35 above)
Comment (sameo) [reply sw/host/opentitantool/src/command/rom_patch.rs]: Oops. Fixed as well, thanks.

File: sw/host/opentitanlib/src/rom_patch/mod.rs (hunk 34 above)
Comment (sameo) [reply sw/host/opentitanlib/src/rom_patch/mod.rs]: That's right. Fixed now.

**Comment by sameo** [comment]
@engdoreis @jesultra Please let me know if this PR needs some more work.

**Comment by jesultra** [comment]
I do not think I can say much more than I already have.  I am not familiar with the feature being built here, but the refactorings in opentitanlib and code overall looks good.  I have not been able to test whether the changes cause issues for `opentitansession` as used at Google, because it seems the branch `integrated_dev` is quite far removed from `master`.  Are these refactorings cherry-picked from `master`, or is there a plan to push the other way?

**Comment by andreaskurth** [comment]
This seems to have gotten plenty of reviews (and it has been open for quite some time), and it looks like all has converged. Big congrats and thanks to everyone involved in this effort!
